controllers/customtasks_controller.go

```diff
@@ -0,0 +1,240 @@
+	run *tknv1alpha1.Run,
+	br *v1alpha1.BuildRun,
+) {
+	if br.Status.CompletionTime != nil {
+		run.Status.CompletionTime = br.Status.CompletionTime
+	}
+	if run.Status.Conditions == nil {
+		run.Status.Conditions = knativev1.Conditions{}
+	}
+
+	var lastCondition apis.Condition
+	if len(run.Status.Conditions) > 0 {
+		lastCondition = run.Status.Conditions[len(run.Status.Conditions)-1]
+	}
+
+	for _, c := range br.Status.Conditions {
+		logger.WithValues(
+			"buildrun-status", c.Status,
+			"buildrun-reason", c.Reason,
+			"buildrun-message", c.Message,
+		).V(0).Info("Reflecting BuildRun status condition on the Tekton Run owner object")
+
+		severity := apis.ConditionSeverityInfo
+		if c.Status == corev1.ConditionFalse {
+			severity = apis.ConditionSeverityError
+		}
+
+		condition := apis.Condition{
+			Type:               apis.ConditionType(string(c.Type)),
+			Status:             c.Status,
+			LastTransitionTime: apis.VolatileTime{Inner: c.LastTransitionTime},
+			Reason:             c.Reason,
+			Message:            c.Message,
+			Severity:           severity,
+		}
+
+		if !reflect.DeepEqual(lastCondition, condition) {
+			run.Status.Conditions = append(run.Status.Conditions, condition)
+		}
+	}
+
+	if len(run.Status.Conditions) == 0 {
+		run.Status.Conditions = []apis.Condition{{
+			Type:               apis.ConditionSucceeded,
+			Status:             corev1.ConditionUnknown,
+			LastTransitionTime: apis.VolatileTime{Inner: metav1.Now()},
+		}}
+	}
+}
+
+// Reconcile reconciles the Custom-Tasks Run instances pointing to Shipwright Builds, by issuing a
+// BuildRun and taking advantage of Status.ExtraFields to record the BuildRun namespace/name. When
+// the ExtraFields is populated the controller uses the reference to reflect the BuildRun status
+// updates on the original Run instance.
+func (r *CustomTasksReconciler) Reconcile(
+	ctx context.Context,
+	req ctrl.Request,
+) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	var run tknv1alpha1.Run
+	err := r.Get(ctx, req.NamespacedName, &run)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			logger.Error(err, "Unable to fetch Run")
+		}
+		return RequeueOnError(client.IgnoreNotFound(err))
+	}
+	originalRun := run.DeepCopy()
+
+	// making sure the current run object status is recorded in the logs
+	logger = logger.WithValues("successful", run.IsSuccessful(), "cancelled", run.IsCancelled())
+
+	// when the run instace is marked as done, no further actions needs to take place
+	if run.IsDone() {
+		logger.V(0).Info("Tekton Run is synchronized, all done!")
+		return Done()
+	}
+
+	// extracting the meta-information recorded in the Tekton Run status, in this section the name of
+	// the BuildRun issued for the object is recorded
+	var extraFields filter.ExtraFields
+	if err := run.Status.DecodeExtraFields(&extraFields); err != nil {
+		logger.V(0).Error(err, "Trying to decode Run Status' ExtraFields")
+		return RequeueOnError(err)
+	}
+
+	var br = &v1alpha1.BuildRun{}
```

Comment:
The logic here is missing code to handle a canceled Run. I suggest the following:

- If extra fields is empty (= BuildRun not yet created), then we do not create a BuildRun at all.
- If extra fields is not empty (= BuildRun was created), then we retrieve the BuildRun, check if it is not complete, if so, we cancel it (= set .spec.status to BuildRunCancel). This will need update permission to be added.

Reply:
Thanks for the suggestion! I added the logic to handle both Tekton Run and BuildRun getting cancelled reflecting the status on each other. Please consider the latest changes.